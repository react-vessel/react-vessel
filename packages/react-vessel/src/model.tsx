import React, { useContext } from 'react';

const ModelContext = React.createContext<string>('');

export function useParentModel(): string {
  return useContext(ModelContext);
}

export const Model: React.FC<{ name: string }> = ({ name, children }) => {
  return <ModelContext.Provider value={name}>{children}</ModelContext.Provider>;
};
