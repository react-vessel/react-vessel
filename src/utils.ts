interface ActionNameParams {
  action: string;
  model?: string;
  vessel: string;
  level?: number;
}

export function complementActionName({
  action,
  model,
  vessel,
  level = 3,
}: ActionNameParams): string {
  let splits = action.split('/');

  if (splits.length > 3) {
    throw new Error(`Given action '${action}' does not match the format: [vessel]/[model]/action`);
  }

  if (splits.length === 1) {
    if (!model) {
      throw new Error(`Could not resolve a model`);
    }
    // action
    splits = [vessel, model, action];
  } else if (splits.length === 2) {
    // model/action
    splits = [vessel, ...splits];
  }

  return splits.slice(3 - level).join('/');
}
