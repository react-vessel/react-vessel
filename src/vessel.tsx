/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useReducer, useCallback, useContext, useMemo } from 'react';
import get from 'lodash.get';
import { ReducerType, Dispatch } from './types';
import { usePersistentMap } from './usePersistentMap';
import { VesselContextType, VesselProps, VesselContext } from './vesselContext';
import { complementActionName } from './utils';
import { useEventBus } from './eventBus';

const RE_REDUCER_KEY = /(.*):(.*)\/(.*)/;

function parseReducerKey(key: string): [string, string] {
  const match = key.match(RE_REDUCER_KEY);

  /**
   * There is only one way to put a reducer into reducer map
   * and reducer key is validated there
   * so typescript should just shut up here
   */
  // @ts-ignore
  return [match[1], `${match[2]}/${match[3]}`];
}

function useVessel(name: string): VesselContextType {
  const [getReducers, setReducer] = usePersistentMap<ReducerType<any, any>>();
  const eventBus = useEventBus();

  const reducer = useCallback<ReducerType<any, any>>((state, action): any => {
    const reducers = getReducers();

    const reducerKeys = Object.keys(reducers);
    const nextState = { ...state };
    let hasChanged = false;

    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < reducerKeys.length; i++) {
      // reducerKey is in the format hostModel:model/action
      const reducerKey = reducerKeys[i];
      const [model, reducerAction] = parseReducerKey(reducerKey);

      if (reducerAction === action.type) {
        const currentReducer = reducers[reducerKey];
        const nextStateForModel = currentReducer(nextState[model], action.payload);
        if (nextStateForModel !== nextState[model]) {
          hasChanged = true;
          nextState[model] = nextStateForModel;
        }
      }
    }

    return hasChanged ? nextState : state;
  }, []);

  const [state, originalDispatch] = useReducer<ReducerType<any, any>>(reducer, {});
  const dispatch = useCallback<Dispatch<any>>(
    (originalAction): void => {
      const actionType = complementActionName({ vessel: name, action: originalAction.type });
      const action = {
        type: actionType,
        payload: originalAction.payload,
      };

      originalDispatch(action);
      eventBus.emit(actionType, action);
    },
    [name, originalDispatch],
  );

  const vessel = useMemo((): VesselContextType => {
    return { name, state, dispatch, setReducer };
  }, [name, state, dispatch, setReducer]);

  return vessel;
}

export function useParentVessel(): VesselContextType {
  return useContext(VesselContext);
}

type StateSelector<T> = (state: any) => T;

export function useParentState<T>(selector: string | StateSelector<T>, defaultValue?: T): T {
  const vessel = useParentVessel();
  if (typeof selector === 'string') {
    return get(vessel.state, selector, defaultValue);
  }

  const value = selector(vessel.state);
  if (typeof defaultValue !== 'undefined' && typeof value === 'undefined') {
    return defaultValue;
  }

  return value;
}

export const Vessel: React.FC<VesselProps> = ({ name = 'default-vessel', children }) => {
  const vessel = useVessel(name);
  return <VesselContext.Provider value={vessel}>{children}</VesselContext.Provider>;
};
