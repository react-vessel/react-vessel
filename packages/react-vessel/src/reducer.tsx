import { useEffect } from 'react';
import { ReducerType } from './types';
import { useStaticCallback } from './useStaticCallback';
import { useParentVessel } from './vessel';
import { complementActionType } from './utils';
import { useParentModel } from './model';

interface ReducerProps {
  model?: string;
  action: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reducer: ReducerType<any, any>;
}

export const Reducer: React.FC<ReducerProps> = ({ model, action, reducer }) => {
  const parentModel = useParentModel();
  const { name: vesselName, setReducer } = useParentVessel();
  const staticReducer = useStaticCallback(reducer);

  const currentModel = model || parentModel;

  const fullActionName = complementActionType({
    action,
    model: currentModel,
    vessel: vesselName,
  });

  useEffect((): (() => void) => {
    const reducerName = `${currentModel}:${fullActionName}`;
    setReducer(reducerName, staticReducer);

    return (): void => {
      setReducer(reducerName, undefined);
    };
  }, [currentModel, fullActionName, staticReducer, setReducer]);

  return null;
};
