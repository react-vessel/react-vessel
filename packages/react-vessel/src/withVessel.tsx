/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { useParentState, useParentVessel } from './vessel';
import { VesselContextType } from './vesselContext';

interface WithVesselProps {
  select: string;
  default?: any;
  render: (selectedState: any, vessel: VesselContextType) => JSX.Element;
}

export const WithVessel: React.FC<WithVesselProps> = ({
  select,
  default: defaultValue,
  render,
}) => {
  const vessel = useParentVessel();
  const selectedState = useParentState(select, defaultValue);
  return render(selectedState, vessel);
};
