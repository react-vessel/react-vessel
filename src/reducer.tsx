import { useEffect } from 'react';
import { ReducerType } from './types';
import { useStaticCallback } from './useStaticCallback';
import { useParentVessel } from './vessel';
import { complementActionName } from './utils';
import { useParentModel } from './model';

interface ReducerProps {
  model?: string;
  action: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reducer: ReducerType<any, any>;
}

export const Reducer: React.FC<ReducerProps> = ({ model, action, reducer }) => {
  const parentModel = useParentModel();
  const vessel = useParentVessel();
  const staticReducer = useStaticCallback(reducer);

  const currentModel = model || parentModel;

  const fullActionName = complementActionName({
    action,
    model: currentModel,
    vessel: vessel.name,
  });
  useEffect((): (() => void) => {
    const reducerName = `${currentModel}:${fullActionName}`;
    vessel.setReducer(reducerName, staticReducer);

    return (): void => {
      vessel.setReducer(reducerName, undefined);
    };
  }, [fullActionName, staticReducer]);

  return null;
};
