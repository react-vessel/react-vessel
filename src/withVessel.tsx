/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { Dispatch } from './types';
import { useParentState, useParentVessel } from './vessel';

interface WithVesselProps {
  select: string;
  default?: any;
  render: (selectedState: any, params: { state: any; dispatch: Dispatch<any> }) => JSX.Element;
}

export const WithVessel: React.FC<WithVesselProps> = ({
  select,
  default: defaultValue,
  render,
}) => {
  const { state, dispatch } = useParentVessel();
  const selectedState = useParentState(select, defaultValue);
  return render(selectedState, { state, dispatch });
};
