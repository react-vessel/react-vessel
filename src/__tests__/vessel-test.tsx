import React, { useState } from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { Reducer } from '../reducer';

import { Vessel, useParentState, useParentVessel } from '../vessel';
import { Model } from '../model';
import { complementActionName } from '../utils';
import { WithVessel } from '../withVessel';

const CounterWithModel: React.FC<{ name: string }> = ({ name }) => {
  return (
    <>
      <Model name={name}>
        <Reducer action="increment" reducer={(state = 0) => state + 1} />
        <Reducer action="decrement" reducer={(state = 0) => state - 1} />
      </Model>
    </>
  );
};

const Counter: React.FC<{ name: string }> = ({ name }) => {
  return (
    <>
      <Reducer model={name} action="increment" reducer={(state = 0) => state + 1} />
      <Reducer model={name} action="decrement" reducer={(state = 0) => state - 1} />
    </>
  );
};

const CounterShow: React.FC<{ name: string }> = ({ name }) => {
  const count = useParentState('count', 0);
  return <div className={`counterShow-${name}`}>{count}</div>;
};

const CounterButtons: React.FC<{ name: string }> = ({ name }) => {
  const vessel = useParentVessel();
  return (
    <div className={`counterButtons-${name}`}>
      <button
        type="button"
        className="increment"
        onClick={() => vessel.dispatch({ type: `${name}/increment` })}
      >
        {`Increment ${name}`}
      </button>
      <button
        type="button"
        className="decrement"
        onClick={() => vessel.dispatch({ type: `${name}/decrement` })}
      >
        {`Decrement ${name}`}
      </button>
    </div>
  );
};

afterEach(cleanup);

describe('Vessel', () => {
  it('renders without a name', () => {
    const { asFragment } = render(<Vessel />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with a name', () => {
    const { asFragment } = render(<Vessel name="test" />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Reducer', () => {
  it('works without a model', () => {
    const { container, getByText } = render(
      <Vessel name="test">
        <Counter name="count" />
        <CounterShow name="count" />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('1');
  });

  it('works with a model', () => {
    const { container, getByText } = render(
      <Vessel name="test">
        <CounterWithModel name="count" />
        <CounterShow name="count" />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('1');
  });

  it('works with a selector method', () => {
    const CounterShowWithSelector: React.FC<{ name: string }> = ({ name }) => {
      const count = useParentState(state => state[name], 0);
      return <div className={`counterShow-${name}`}>{count}</div>;
    };

    const { container, getByText } = render(
      <Vessel name="test">
        <CounterShowWithSelector name="count" />
        <CounterWithModel name="count" />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('1');
  });

  it('unmounts cleanly', () => {
    function CounterWithToggle() {
      const [rendered, setRendered] = useState(true);

      return (
        <>
          {rendered && <Counter name="count" />}

          <button type="button" onClick={() => setRendered(!rendered)}>
            Toggle Counter
          </button>
        </>
      );
    }

    const { container, getByText } = render(
      <Vessel name="test">
        <CounterWithToggle />
        <CounterShow name="count" />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('1');

    // unmount the counter
    fireEvent.click(getByText('Toggle Counter'));

    // count should still be '1' because we unmounted the counter
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.counterShow-count`)!.textContent).toEqual('1');
  });

  describe('', () => {
    beforeEach(() => {
      jest.spyOn(console, 'error');
      // @ts-ignore
      console.error.mockImplementation(() => {});
    });

    afterEach(() => {
      // @ts-ignore
      console.error.mockRestore();
    });

    it('throws with wrong action types', () => {
      function IncorrectCounter() {
        return (
          <>
            <Reducer model="count" action="a/b/c/d" reducer={state => state} />
          </>
        );
      }

      expect(() => {
        render(
          <Vessel name="test">
            <IncorrectCounter />
          </Vessel>,
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('throws if cannot resolve a model', () => {
      function IncorrectCounter() {
        return (
          <>
            <Reducer action="increment" reducer={state => state} />
          </>
        );
      }

      expect(() => {
        render(
          <Vessel name="test">
            <IncorrectCounter />
          </Vessel>,
        );
      }).toThrowErrorMatchingSnapshot();
    });
  });
});

test('complementActionName works with different levels', () => {
  expect(complementActionName({ action: 'increment', model: 'count', vessel: 'vessel' })).toEqual(
    'vessel/count/increment',
  );

  expect(
    complementActionName({ action: 'increment', model: 'count', vessel: 'vessel', level: 2 }),
  ).toEqual('count/increment');

  expect(
    complementActionName({ action: 'increment', model: 'count', vessel: 'vessel', level: 1 }),
  ).toEqual('increment');
});

describe('WithVessel', () => {
  it('works without default value', () => {
    const { container, getByText } = render(
      <Vessel name="test">
        <CounterWithModel name="count" />
        <WithVessel
          select="count"
          render={count => {
            return <div className="count-element">{count}</div>;
          }}
        />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.count-element`)!.textContent).toEqual('');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.count-element`)!.textContent).toEqual('1');
  });

  it('works with default value', () => {
    const { container, getByText } = render(
      <Vessel name="test">
        <CounterWithModel name="count" />
        <WithVessel
          select="count"
          default={0}
          render={count => {
            return <div className="count-element">{count}</div>;
          }}
        />
        <CounterButtons name="count" />
      </Vessel>,
    );

    expect(container.querySelector(`.count-element`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.count-element`)!.textContent).toEqual('1');
  });

  it('brings dispatch', () => {
    const { container, getByText } = render(
      <Vessel name="test">
        <CounterWithModel name="count" />
        <WithVessel
          select="count"
          default={0}
          render={(count, { dispatch }) => {
            return (
              <>
                <div className="count-element">{count}</div>
                <button type="button" onClick={() => dispatch({ type: 'count/increment' })}>
                  Increment count
                </button>
                <button type="button" onClick={() => dispatch({ type: 'count/decrement' })}>
                  Decrement count
                </button>
              </>
            );
          }}
        />
      </Vessel>,
    );

    expect(container.querySelector(`.count-element`)!.textContent).toEqual('0');
    fireEvent.click(getByText('Increment count'));
    expect(container.querySelector(`.count-element`)!.textContent).toEqual('1');
  });
});
