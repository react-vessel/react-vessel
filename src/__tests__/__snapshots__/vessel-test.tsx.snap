// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer  throws if cannot resolve a model 1`] = `"Could not resolve a model"`;

exports[`Reducer  throws with wrong action types 1`] = `"Given action 'a/b/c/d' does not match the format: [vessel]/[model]/action"`;

exports[`Vessel renders with a name 1`] = `<DocumentFragment />`;

exports[`Vessel renders without a name 1`] = `<DocumentFragment />`;
